### PR TITLE
MCOL-5467 Add support for duplicate expressions in group by.

### PR DESCRIPTION
**Pull request by denis0x0D**
This patch adds support for duplicate expressions (builtin_functions) with one argument in select statement and group by statement.

### CODE REVIEW

File: dbcon/joblist/jlf_common.h [hunk 1]
```diff
@@ -93,6 +93,15 @@ struct TupleInfo
   uint32_t csNum;  // For collations
 };
 
+// This struct holds information about `FunctionColumn`.
+struct FunctionColumnInfo
+{
+  // Function argument.
+  uint64_t associatedColumnOid;
+  // Function name.
+  std::string functionName;
```
Comment (drrtuy) [inline dbcon/joblist/jlf_common.h:102]: Mb ref?

File: dbcon/joblist/joblistfactory.cpp [hunk 2]
```diff
@@ -1246,6 +1248,7 @@ const JobStepVector doAggProject(const CalpontSelectExecutionPlan* csep, JobInfo
         const WindowFunctionColumn* wc = NULL;
         bool hasAggCols = false;
         bool hasWndCols = false;
+        bool hasFuncCols = false;
```
Comment (drrtuy) [inline dbcon/joblist/joblistfactory.cpp]: It is better to rename the flag so it brings the specific semantics, e.g. hasFuncColsWithSingleArg

File: dbcon/joblist/joblistfactory.cpp [hunk 3]
```diff
@@ -1291,6 +1297,14 @@ const JobStepVector doAggProject(const CalpontSelectExecutionPlan* csep, JobInfo
         {
           jobInfo.expressionVec.push_back(tupleKey);
         }
+
+        if (hasFuncCols)
+        {
+          FunctionColumnInfo fcInfo;
```
Comment (drrtuy) [inline dbcon/joblist/joblistfactory.cpp]: You should better add a proper ctor.

File: mysql-test/columnstore/bugfixes/mcol-5476.result [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+insert into t1 values (1, 1), (2, 1), (3, 1), (4, 2), (5, 2);
+select sum(a), abs(b), abs(b) from t1 group by abs(b), abs(b) order by abs(b);
+sum(a)	abs(b)	abs(b)
+6	1	1
+9	2	2
+select sum(a), abs(b), abs(b) from t1 group by abs(b) order by abs(b);
+sum(a)	abs(b)	abs(b)
+6	1	1
+9	2	2
+select sum(distinct a), abs(b), abs(b) from t1 group by abs(b), abs(b) order by abs(b);
```
Comment (drrtuy) [inline mysql-test/columnstore/bugfixes/mcol-5476.result]: You should use #sorted_result MTR interpreter clause. This way MTR sorts the given and the expected outputs on its own before comparison. You can drop ORDER BY using #sorted_result.

Comment (drrtuy) [reply mysql-test/columnstore/bugfixes/mcol-5476.result]: This is valid for all similar cases in this test.

File: dbcon/joblist/tupleaggregatestep.cpp [hunk 5]
```diff
@@ -5953,4 +5984,42 @@ void TupleAggregateStep::formatMiniStats()
...
+uint32_t TupleAggregateStep::getTupleKeyFromTuple(uint32_t key)
+{
+  return key;
+}
+
+template <class GroupByMap>
+bool TupleAggregateStep::tryToFindEqualFunctionColumnByTupleKey(JobInfo& jobInfo, GroupByMap& groupByMap,
+                                                                const uint32_t tupleKey, uint32_t& foundKey)
+{
+  if (jobInfo.functionColumnMap.count(tupleKey))
```
Comment (drrtuy) [inline dbcon/joblist/tupleaggregatestep.cpp]: Change to find/cmp with map::end() as per offline discussion.

File: dbcon/joblist/jlf_common.h (hunk 1 above)
Comment (denis0x0D) [reply dbcon/joblist/jlf_common.h:102]: I think it's ok to hold a copy of string, because I'm not sure when FunctionColumn destructor is about to call.

**Review by drrtuy — APPROVED** [review]

**Comment by drrtuy** [comment]
untested ACK

**Pull request merged**